### ISSUE TITLE
[Question] Select enclosing brackets when cursor is inside a different type of bracket

### ISSUE THREAD

**Issue by acheronfail** [facelessuser/BracketHighlighter#492]
I've been looking through the documentation and all the GH issues, but I haven't found a solution to my problem (sorry if such an issue exists and I've missed it).

## Description

I have a question regarding how to create some custom commands. I'd like to create a shortcut that can select between brackets, even if other brackets exist between them.

I think this is best explained with an example:

```js
[
	// BEFORE:
	{ "keys":               ["super+k", "a"], "command": "my_command" },
	// Cursor inside string:    ^

	// AFTER:
	{ "keys":               ["super+k", "a"], "command": "my_command" },
	// Cursor wraps content: ^------------^
]
```

With the cursor inside the string, I'd like to create a shortcut that will select everything between the enclosing square brackets. At the moment, this doesn't work (I think BH sees that it's inside a `"` set of brackets, and bails out of the command).

This is my current command:

```json
{ 
    "keys": ["super+k", "["],
    "command": "bh_key",
    "args": {
        "lines" : true,
        "plugin": {
            "type": ["square"],
            "command": "bh_modules.bracketselect"
        }
    } 
},
```

So, this shortcut only works if the cursor *is not already between* a different type of bracket. I'm fairly sure there's a way to do this with BH, but I haven't yet figured it out yet.

## Support Info

```text
- ST ver.: 3175
- Platform: osx
- Arch: x64
- Plugin ver.: 2.27.4
- Install via PC: True
- mdpopups ver.: 3.3.3
- backrefs ver.: 3.5
- markdown ver.: 2.6.11
- pygments ver.: 2.1a0
- jinja2 ver.: 2.8
```

My question is, how can I configure BH so that it selects enclosing/wrapping brackets, even if the cursor is currently within a *different* set of brackets?

**Comment by facelessuser** [comment]
You'll have to give me some time to get around to looking at this. I'm not sure I understand exactly what the issue is, but I'll hopefully get around to looking at this soon.

**Comment by facelessuser** [comment]
So, as far as I can tell, the command below works fine.  At least as I would expect it to work.  I will elaborate.

```js
    // Select text between brackets
    {
        "keys": ["super+k", "["],
        "command": "bh_key",
        "args":
        {
            "lines" : true,
            "plugin":
            {
                "type": ["square"],
                "command": "bh_modules.bracketselect"
            }
        }
    },
```

BH, no matter what, targets the **closest**, valid bracket around a cursor(s), and only that **one** (per cursor).  Then the plugin is applied to **that** target if it meets the appropriate criteria.

BH does **not** match **all** brackets around the cursor and then apply the plugin to one of the **many** matched brackets.

Setting type `square` does **not** currently filter BH's actual matching rules, it only controls when the plugin is applied to a given matched target.

Hopefully that makes it a little clearer?

**Comment by acheronfail** [comment]
> BH, no matter what, targets the closest, valid bracket around a cursor(s), and only that one (per cursor). Then the plugin is applied to that target if it meets the appropriate criteria.

Hmm, right. I suppose then, the behaviour I'm looking for is not something that BH actually does at the moment?

To be really clear, I've attached a gif of what I'd like to do. Take note, that every single command I run in the gif should be to select `square` brackets.

![selection](https://user-images.githubusercontent.com/34289614/43816315-5d55ddc8-9b17-11e8-8c37-fe4ce28f30db.gif)

So, what I'm looking for, is to create a shortcut where BH will select the nearest pair of `square` brackets from my cursor (ignoring other brackets on the way).

Am I correct in thinking this isn't currently possible?

**Comment by facelessuser** [comment]
> So, what I'm looking for, is to create a shortcut where BH will select the nearest pair of square brackets from my cursor (ignoring other brackets on the way).

If the closest surrounding bracket is a square, then it will select the content of the square, but if the closest bracket is something else, it should cancel the command.

The only way to get what you want would to alter the list of valid bracket patterns on each command instance on the fly. Like if I set `type` to `square`, BH only looks for squares and no other bracket types.  You lose context in that case so something like `([)(])` would select the `[)(]` which normally wouldn't get selected due to context with `()` patterns.  Or you would have to keep matching brackets further and further out until you find a valid one.  Performance on executions could be quite bad in certain cases.

Basically, out of the box, BH matches the closest bracket and immediately stops searching for performance reasons.  No parent brackets are matched.  There is no current way to match anything except the current surrounding bracket, but you conditionally due things based on the type of surrounding bracket.

So, in short, the answer is no.

**Comment by acheronfail** [comment]
> Basically, out of the box, BH matches the closest bracket and immediately stops searching for performance reasons. No parent brackets are matched. There is no current way to match anything except the current surrounding bracket, but you conditionally due things based on the type of surrounding bracket.
> 
> So, in short, the answer is no.

Yeah, I figured something like this would be the case.

I appreciate the time you've taken to talk to me about this! And I'm not sure the benefits of adding this feature to BH outweigh complexity and performance concerns. 😄

**Comment by acheronfail** [comment]
@facelessuser FWIW, I managed to get the functionality I wanted by using a custom run instance plugin.

You'll probably be ashamed of the performance considerations of this code 😅 haha, but I only wanted this for a way to quickly select a specific type of enclosing brackets. It works well enough for me at the moment, I'll test it out and see how I go.

Here it is:

```python
import BracketHighlighter.bh_plugin as bh_plugin
import sublime

# How many characters to search through before giving up.
SEARCH_THRESHOLD = 5000

class NaiveBracketSelect(bh_plugin.BracketPluginCommand):
  """
  A naïve bracket selector. This plugin will search for the nearest enclosing
  pair of brackets that you pass into it, whether or not the cursor is already
  inside a different type of bracket.
  """

  def _find(self, start, args, region, is_left):
    """Find the nearest bracket before/after the starting position."""

    balance = 0
    target = args['left' if is_left else 'right']
    other = args['right' if is_left else 'left']
    text = self.view.substr(region)

    for i, ch in enumerate(reversed(text) if is_left else text):
      if balance == 0 and ch == target:
        return start - i if is_left else start + i

      if args['balanced']:
        if ch == other:
          balance += 1
        if ch == target:
          balance -= 1

    return None

  def _left_region(self, start):
    """Max region of text to search to the left."""
    return sublime.Region(max(0, start - SEARCH_THRESHOLD), start)

  def _right_region(self, start):
    """Max region of text to search to the right."""
    return sublime.Region(start, min(self.view.size(), start + SEARCH_THRESHOLD))

  def run(self, edit, name, **kwargs):
    new_selection = []
    for sel in self.selection:
      a, b = sel.begin(), sel.end()
      left = self._find(a, kwargs, self._left_region(a), True)
      right = self._find(b, kwargs, self._right_region(b), False)

      # Do nothing if we couldn't find both brackets.
      if left == None or right == None:
        return

      # If we're already sitting on a pair of brackets, expand to select them.
      if left == a and right == b:
        new_selection.append(sublime.Region(left - 1, right + 1))
      else:
        new_selection.append(sublime.Region(left, right))

    self.selection = new_selection


def plugin():
  return NaiveBracketSelect
```

And the key binding:

```js
{
    "keys": ["super+k", "["],
    "command": "bh_key",
    "args": {
        "lines" : true,
        "plugin": {
            "type": ["__all__"],
            "command": "User.bh_modules.naivebracketselect",
            "args": {
                "left": "[",
                "right": "]",
                "balanced": true
            }
        }
    }
}
```

**Comment by facelessuser** [comment]
Well, I'm glad you found something to get what you need 😄 .

There are a number of features that I think would be cool, but really are more practical if implemented (or at least partially implemented) on a much lower level to get good support, which isn't really possible with the current Sublime API.

More complexity could be integrated into the source to maybe allow cooler things, but honestly, maintainability is my biggest priority at this point.  I maintain too many plugins and repos.  Most of my plugins are more in a state of fixing bugs and maintaining support with the latest ST and only the occasional new feature.  That's not to say I don't add new features, but I am a lot more selective these days.
